core/Pos/Binary/Class/Primitive.hs

```diff
@@ -3,110 +3,135 @@
-import           Pos.Binary.Class.Numbers   (TinyVarInt (..), UnsignedVarInt (..),
-                                             getWord8, putWord8)
-import           Pos.Binary.Class.Store     (Peek, Poke, PokeWithSize (..), Size (..),
-                                             StaticSize (..), putS)
-
-import           Pos.Binary.Class.Instances ()
--- TODO ^ Move Raw in Pos.Util.Util and remove the import.
-
--- | See 'putWord8'.
-putWord8S :: Word8 -> PokeWithSize ()
-putWord8S = putS @Word8
-
--- | See 'putBytes'.
-putBytesS :: ByteString -> PokeWithSize ()
-putBytesS bs = PokeWithSize (BS.length bs) (putBytes bs)
+-- | Serialize a Haskell value to an external binary representation.
+--
+-- The output is represented as a lazy 'BSL.ByteString' and is constructed
+-- incrementally.
+serialize :: Bi a => a -> BSL.ByteString
+serialize = CBOR.Write.toLazyByteString . encode
 
-----------------------------------------------------------------------------
--- Raw
-----------------------------------------------------------------------------
+-- | Serialize a Haskell value to an external binary representation.
+--
+-- The output is represented as a strict 'BS.ByteString'.
+serialize' :: Bi a => a -> BS.ByteString
+serialize' = BSL.toStrict . serialize
 
--- | A wrapper over 'ByteString' for signalling that a bytestring should be
--- processed as a sequence of bytes, not as a separate “entity”. It's used in
--- crypto and binary code.
-newtype Raw = Raw ByteString
-    deriving (Bi, Eq, Ord, Show, Typeable, NFData)
+-- | Deserialize a Haskell value from the external binary representation
+-- (which must have been made using 'serialize' or related function).
+--
+-- /Throws/: @'CBOR.Read.DeserialiseFailure'@ if the given external
+-- representation is invalid or does not correspond to a value of the
+-- expected type.
+deserialize :: Bi a => BSL.ByteString -> a
+deserialize bs0 =
+    runST (supplyAllInput bs0 =<< deserializeIncremental)
+  where
+    supplyAllInput _bs (CBOR.Read.Done _ _ x) = return x
+    supplyAllInput  bs (CBOR.Read.Partial k)  =
+      case bs of
+        BSL.Chunk chunk bs' -> k (Just chunk) >>= supplyAllInput bs'
+        BSL.Empty           -> k Nothing      >>= supplyAllInput BSL.Empty
+    supplyAllInput _ (CBOR.Read.Fail _ _ exn) = throw exn
+
+-- | Strict variant of 'deserialize'.
+deserialize' :: Bi a => BS.ByteString -> a
+deserialize' = deserialize . BSL.fromStrict
+
+decodeFull :: Bi a => BS.ByteString -> Either Text a
+decodeFull bs0 = case deserializeOrFail' bs0 of
+  Right x -> pure x
+  Left  e -> Left $ fromString (show e)
+
+-- | Deserialize a Haskell value from the external binary representation,
+-- or get back a @'DeserialiseFailure'@.
+deserializeOrFail :: Bi a => BSL.ByteString -> Either CBOR.Read.DeserialiseFailure a
+deserializeOrFail bs0 =
+    runST (supplyAllInput bs0 =<< deserializeIncremental)
+  where
+    supplyAllInput _bs (CBOR.Read.Done _ _ x) = return (Right x)
```

Comment:
You commented elsewhere about leftovers. Here is where we would check for leftovers. Depending on how we're doing framing then checking this would be critical for ensuring a canonical form.